### PR TITLE
Skyline: adding "SkyLint", a simple program that checks for code issues that ReSharper doesn't handle to our satisfaction

### PR DESCRIPTION
**Pull request by bspratt**
In particular, inspections in generated code which ReSharper ignores.

So why isn't this handled with git pre-commit hooks? A few reasons - GitHub doesn't support server-side hooks (as a security issue) and client side there's no agreement on how commit hooks are handled between commandline git and TortoiseGit. Also, it's nice to be told about the problem while you're working on the code rather than just as you think you're done with it.

So, this little app is added as a Skyline pre-build step, it throws up a messagebox if it sees a prohibited string in any file matching a file mask, then exits with a non-zero code.

**Note that this commit is actually an intentionally broken build**, as there are a couple of lingering examples of ".ImageScalingSize = new System.Drawing.Size(" in the code, which this whole project is designed to ferret out.  That will need to be corrected on merge to master.

### CODE REVIEW

**Comment by bspratt** [comment]
Setting this to "Draft" because I don't want it to hang up on TeamCity

**Comment by nickshulman** [comment]
Would it have been easier to write a unit test which searches for that string in the source tree?
I know there are some cases where the source tree is not available (like when you download SkylineTester.zip), but the unit test could check whether that is the case, and skip checking anything.
But, the test would work in the cases we care about, which would be TeamCity and when developers run tests.

**Comment by bspratt** [comment]
That's a great idea - I sort of wandered into this implementation after wrestling with git hooks for a while, a unit test is simpler. Thanks!

**Comment by brendanx67** [comment]
Totally agree. We want our tests to catch things as much as possible. Clearly it was a mistake to think ReSharper notifications would work on TeamCity and had the potential to notify a developer as soon as the problem was added to the project, before tests got run. Just not the way we work and not supported by our TeamCity code inspection.

**Comment by chambm** [comment]
Is it desirable to only run tests if some linting/inspection or special test passes? That could be handled by the smartTrigger python script.

**Comment by bspratt** [comment]
> Is it desirable to only run tests if some linting/inspection or special test passes? That could be handled by the smartTrigger python script.

I think what we have here suits us for now, and it alerts the developer well before the changes make it to TeamCity, which is ideal.

**Comment by bspratt** [comment]
> I know there are some cases where the source tree is not available (like when you download SkylineTester.zip), but the unit test could check whether that is the case, and skip checking anything.

I haven't covered that case - but I think a lot of other tests would fail in that scenario too as their data sets reside in the source tree.

**Comment by bspratt** [comment]
Note that this is failing on TeamCity [https://teamcity.labkey.org/viewLog.html?buildId=1112619&tab=buildResultsDiv&buildTypeId=bt19](https://teamcity.labkey.org/viewLog.html?buildId=1112619&tab=buildResultsDiv&buildTypeId=bt19)  which is a good thing. I'll now go and remove the last two offending occurrences of `".ImageScalingSize = new System.Drawing.Size("` in the branch and it should go green.

**Review by nickshulman — COMMENTED** [review]
Looks great to me.

I was planning on writing a test which makes sure that all the entries in ColumnCaptions.resx are in alphabetical order. It would not really fit into this framework here, but it could leverage whatever logic you have for finding the location of the source tree. Maybe that logic could go somewhere in TestUtil.

File: pwiz_tools/Skyline/Test/CodeInspectionTest.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using Microsoft.VisualStudio.TestTools.UnitTesting;
+using pwiz.SkylineTestUtil;
+
+namespace pwiz.SkylineTest
+{
+    /// <summary>
+    /// Summary description for RefineTest
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Test/CodeInspectionTest.cs:37]: nice catch thanks

**Comment by bspratt** [comment]
I'm a little nervous ripping those offending lines out since I didn't put them in - this is just undesirable boilerplate from the VisualStudio designer, I take it?

File: pwiz_tools/Skyline/Test/CodeInspectionTest.cs [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    public class CodeInspectionTest : AbstractUnitTest
+    {
+        private static Dictionary<string, Dictionary<string, string>> forbidden = new Dictionary<string, Dictionary<string, string>>
+        {
+            { @"*.Designer.cs", // File types to inspect for this error
+                new Dictionary<string, string>
+                {
+                    {
+                        @".ImageScalingSize = new System.Drawing.Size(", // Forbidden pattern
+                        @"causes issues on HD monitors" // Explanation for prohibition
```
[the remark this answers is not in the document]
Comment (bspratt) [reply pwiz_tools/Skyline/Test/CodeInspectionTest.cs:50]: > Maybe this should say "causes icons to be blurry".
> I don't think the problems are specific to HD monitors.

I was just using the text Brendan had in the ReSharper custom inspection, but I can combine them.

**Pull request merged**